In [372]:
import pandas as pd

In [373]:
mpg_df = pd.read_csv('C:/ex/data/archive (1)/auto-mpg - 복사본.csv')

In [374]:
mpg_df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0        130    3504          12.0   
1    15.0          8         350.0        165    3693          11.5   
2    18.0          8         318.0        150    3436          11.0   
3    16.0          8         304.0        150    3433          12.0   
4    17.0          8         302.0        140    3449          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0         86    2790          15.6   
394  44.0          4          97.0         52    2130          24.6   
395  32.0          4         135.0         84    2295          11.6   
396  28.0          4         120.0         79    2625          18.6   
397  31.0          4         119.0         82    2720          19.4   

     model year  origin                   car name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[398 rows x 9 columns]

In [375]:
mpg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.1+ KB


In [376]:
# 398*9

In [377]:
# 1 mpg = 0.425143707 km/l

In [378]:
mpg_df['mpg']*0.425143707

0       7.652587
1       6.377156
2       7.652587
3       6.802299
4       7.227443
         ...    
393    11.478880
394    18.706323
395    13.604599
396    11.904024
397    13.179455
Name: mpg, Length: 398, dtype: float64

In [379]:
mpg_df['kpl'] = round(mpg_df['mpg']*0.425143707,2)

In [380]:
mpg_df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name   kpl  
0       1  chevrolet chevelle malibu  7.65  
1       1          buick skylark 320  6.38  
2       1         plymouth satellite  7.65  
3       1              amc rebel sst  6.80  
4       1                ford torino  7.23

In [381]:
mpg_df.horsepower.dtype

dtype('O')

In [382]:
mpg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
 9   kpl           398 non-null    float64
dtypes: float64(4), int64(4), object(2)
memory usage: 31.2+ KB


In [383]:
mpg_df['horsepower'].unique()

array(['130', '165', '150', '140', '198', '220', '215', '225', '190',
       '170', '160', '95', '97', '85', '88', '46', '87', '90', '113',
       '200', '210', '193', '?', '100', '105', '175', '153', '180', '110',
       '72', '86', '70', '76', '65', '69', '60', '80', '54', '208', '155',
       '112', '92', '145', '137', '158', '167', '94', '107', '230', '49',
       '75', '91', '122', '67', '83', '78', '52', '61', '93', '148',
       '129', '96', '71', '98', '115', '53', '81', '79', '120', '152',
       '102', '108', '68', '58', '149', '89', '63', '48', '66', '139',
       '103', '125', '133', '138', '135', '142', '77', '62', '132', '84',
       '64', '74', '116', '82'], dtype=object)

In [384]:
mpg_df[mpg_df['horsepower'] == '?']

mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?    2046          19.0   
126  21.0          6         200.0          ?    2875          17.0   
330  40.9          4          85.0          ?    1835          17.3   
336  23.6          4         140.0          ?    2905          14.3   
354  34.5          4         100.0          ?    2320          15.8   
374  23.0          4         151.0          ?    3035          20.5   

     model year  origin              car name    kpl  
32           71       1            ford pinto  10.63  
126          74       1         ford maverick   8.93  
330          80       2  renault lecar deluxe  17.39  
336          80       1    ford mustang cobra  10.03  
354          81       2           renault 18i  14.67  
374          82       1        amc concord dl   9.78

In [385]:
import numpy as np

In [386]:
mpg_df['horsepower'].replace('?',np.nan,inplace=True)

C:\Users\acy\AppData\Local\Temp\ipykernel_5092\1287491213.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  mpg_df['horsepower'].replace('?',np.nan,inplace=True)


In [387]:
mpg_df[mpg_df['horsepower'] == '?']

Empty DataFrame
Columns: [mpg, cylinders, displacement, horsepower, weight, acceleration, model year, origin, car name, kpl]
Index: []

In [388]:
mpg_df['horsepower'].unique()

array(['130', '165', '150', '140', '198', '220', '215', '225', '190',
       '170', '160', '95', '97', '85', '88', '46', '87', '90', '113',
       '200', '210', '193', nan, '100', '105', '175', '153', '180', '110',
       '72', '86', '70', '76', '65', '69', '60', '80', '54', '208', '155',
       '112', '92', '145', '137', '158', '167', '94', '107', '230', '49',
       '75', '91', '122', '67', '83', '78', '52', '61', '93', '148',
       '129', '96', '71', '98', '115', '53', '81', '79', '120', '152',
       '102', '108', '68', '58', '149', '89', '63', '48', '66', '139',
       '103', '125', '133', '138', '135', '142', '77', '62', '132', '84',
       '64', '74', '116', '82'], dtype=object)

In [389]:
mpg_df['horsepower'] = mpg_df['horsepower'].astype(float)

In [390]:
mpg_df['horsepower'].dtype

dtype('float64')

In [391]:
mpg_df['horsepower'].mean()

104.46938775510205

In [392]:
mpg_df['origin'] #1은 usa, 2는 eu, 3은 japan

0      1
1      1
2      1
3      1
4      1
      ..
393    1
394    2
395    1
396    1
397    1
Name: origin, Length: 398, dtype: int64

In [393]:
mpg_df['origin'].value_counts(dropna=False)

origin
1    249
3     79
2     70
Name: count, dtype: int64

In [394]:
#replace를 이용해서 1,2,3을 해당 국가명으로 바꾼다

In [395]:
mpg_df['origin'].replace({1:'USA',2:'EU',3:'JAPAN'},inplace=True)

In [396]:
mpg_df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year origin                   car name    kpl  
0            70    USA  chevrolet chevelle malibu   7.65  
1            70    USA          buick skylark 320   6.38  
2            70    USA         plymouth satellite   7.65  
3            70    USA              amc rebel sst   6.80  
4            70    USA                ford torino   7.23  
..          ...    ...                        ...    ...  
393          82    USA            ford mustang gl  11.48  
394          82     EU                  vw pickup  18.71  
395          82    USA              dodge rampage  13.60  
396          82    USA                ford ranger  11.90  
397          82    USA                 chevy s-10  13.18  

[398 rows x 10 columns]

In [397]:
mpg_df['origin'] = mpg_df['origin'].astype('category')
# 범주형 데이터로 변환

In [398]:
mpg_df['origin'].dtype

CategoricalDtype(categories=['EU', 'JAPAN', 'USA'], ordered=False, categories_dtype=object)

In [399]:
mpg_df['origin'] = mpg_df['origin'].astype('str')

In [400]:
mpg_df['origin'].dtype

dtype('O')

In [401]:
mpg_df.columns

Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'model year', 'origin', 'car name', 'kpl'],
      dtype='object')

In [402]:
mpg_df['model year']

0      70
1      70
2      70
3      70
4      70
       ..
393    82
394    82
395    82
396    82
397    82
Name: model year, Length: 398, dtype: int64

In [403]:
mpg_df['model year'].sample()

314    80
Name: model year, dtype: int64

In [404]:
mpg_df['model year'].sample(2)

233    77
272    78
Name: model year, dtype: int64

In [405]:
mpg_df['horsepower']

0      130.0
1      165.0
2      150.0
3      150.0
4      140.0
       ...  
393     86.0
394     52.0
395     84.0
396     79.0
397     82.0
Name: horsepower, Length: 398, dtype: float64

In [406]:
mpg_df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year origin                   car name    kpl  
0            70    USA  chevrolet chevelle malibu   7.65  
1            70    USA          buick skylark 320   6.38  
2            70    USA         plymouth satellite   7.65  
3            70    USA              amc rebel sst   6.80  
4            70    USA                ford torino   7.23  
..          ...    ...                        ...    ...  
393          82    USA            ford mustang gl  11.48  
394          82     EU                  vw pickup  18.71  
395          82    USA              dodge rampage  13.60  
396          82    USA                ford ranger  11.90  
397          82    USA                 chevy s-10  13.18  

[398 rows x 10 columns]

In [407]:
mpg_df['horsepower'].mean()

104.46938775510205

In [408]:
horsepower_mean = round(mpg_df['horsepower'].mean(),0)
horsepower_mean

104.0

In [409]:
mpg_df['horsepower'].replace(np.nan,horsepower_mean)

0      130.0
1      165.0
2      150.0
3      150.0
4      140.0
       ...  
393     86.0
394     52.0
395     84.0
396     79.0
397     82.0
Name: horsepower, Length: 398, dtype: float64

In [410]:
mpg_df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year origin                   car name    kpl  
0            70    USA  chevrolet chevelle malibu   7.65  
1            70    USA          buick skylark 320   6.38  
2            70    USA         plymouth satellite   7.65  
3            70    USA              amc rebel sst   6.80  
4            70    USA                ford torino   7.23  
..          ...    ...                        ...    ...  
393          82    USA            ford mustang gl  11.48  
394          82     EU                  vw pickup  18.71  
395          82    USA              dodge rampage  13.60  
396          82    USA                ford ranger  11.90  
397          82    USA                 chevy s-10  13.18  

[398 rows x 10 columns]

In [411]:
mpg_df[mpg_df['horsepower'] == horsepower_mean]

Empty DataFrame
Columns: [mpg, cylinders, displacement, horsepower, weight, acceleration, model year, origin, car name, kpl]
Index: []

In [412]:
mpg_df['horsepower'].max()

230.0

In [413]:
mpg_df['horsepower'].min()

46.0

In [414]:
(mpg_df['horsepower'].max()-mpg_df['horsepower'].min())/3

61.333333333333336

In [415]:
interval = 61.333333333333336

In [416]:
low = mpg_df['horsepower'].min() + interval
medium = low+interval
high = medium+interval

In [417]:
medium

168.66666666666669

In [418]:
print(f'46.0~{low}:저출력')
print(f'{low}~{medium}:보통출력')
print(f'{medium}~{high}:고출력')

46.0~107.33333333333334:저출력
107.33333333333334~168.66666666666669:보통출력
168.66666666666669~230.00000000000003:고출력


In [419]:
mpg_df['horsepower'].values

array([130., 165., 150., 150., 140., 198., 220., 215., 225., 190., 170.,
       160., 150., 225.,  95.,  95.,  97.,  85.,  88.,  46.,  87.,  90.,
        95., 113.,  90., 215., 200., 210., 193.,  88.,  90.,  95.,  nan,
       100., 105., 100.,  88., 100., 165., 175., 153., 150., 180., 170.,
       175., 110.,  72., 100.,  88.,  86.,  90.,  70.,  76.,  65.,  69.,
        60.,  70.,  95.,  80.,  54.,  90.,  86., 165., 175., 150., 153.,
       150., 208., 155., 160., 190.,  97., 150., 130., 140., 150., 112.,
        76.,  87.,  69.,  86.,  92.,  97.,  80.,  88., 175., 150., 145.,
       137., 150., 198., 150., 158., 150., 215., 225., 175., 105., 100.,
       100.,  88.,  95.,  46., 150., 167., 170., 180., 100.,  88.,  72.,
        94.,  90.,  85., 107.,  90., 145., 230.,  49.,  75.,  91., 112.,
       150., 110., 122., 180.,  95.,  nan, 100., 100.,  67.,  80.,  65.,
        75., 100., 110., 105., 140., 150., 150., 140., 150.,  83.,  67.,
        78.,  52.,  61.,  75.,  75.,  75.,  97.,  9

46.0~107.33333333333334:저출력

107.33333333333334~168.66666666666669:보통출력

168.66666666666669~230.00000000000003:고출력

# 저,보통,고출력 기준으로 구간 나누기

In [420]:
bins = [46.0,low,medium,high]
labels = ['저출력','보통출력','고출력']

In [421]:
mpg_df['section of horsepower']= pd.cut(mpg_df['horsepower'],bins=bins,labels=labels,right=True)
#오른쪽 경계값을 포함할려면 right=true

In [422]:
mpg_df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year origin                   car name    kpl section of horsepower  
0            70    USA  chevrolet chevelle malibu   7.65                  보통출력  
1            70    USA          buick skylark 320   6.38                  보통출력  
2            70    USA         plymouth satellite   7.65                  보통출력  
3            70    USA              amc rebel sst   6.80                  보통출력  
4            70    USA                ford torino   7.23                  보통출력  
..          ...    ...                        ...    ...                   ...  
393          82    USA            ford mustang gl  11.48                   저출력  
394          82     EU                  vw pickup  18.71                   저출력  
395          82    USA              dodge rampage  13.60                   저출력  
396          82    USA                ford ranger  11.90                   저출력  
397          82    USA                 chevy s-10  13.18                   저출력  

[398 rows x 11 columns]

In [423]:
mpg_df.insert(mpg_df.columns.get_loc('horsepower') + 1, 'section of horsepower', mpg_df.pop('section of horsepower'))
# 'horsepower' 옆에 'category' 칼럼을 삽입

In [424]:
mpg_df

mpg  cylinders  displacement  horsepower section of horsepower  weight  \
0    18.0          8         307.0       130.0                  보통출력    3504   
1    15.0          8         350.0       165.0                  보통출력    3693   
2    18.0          8         318.0       150.0                  보통출력    3436   
3    16.0          8         304.0       150.0                  보통출력    3433   
4    17.0          8         302.0       140.0                  보통출력    3449   
..    ...        ...           ...         ...                   ...     ...   
393  27.0          4         140.0        86.0                   저출력    2790   
394  44.0          4          97.0        52.0                   저출력    2130   
395  32.0          4         135.0        84.0                   저출력    2295   
396  28.0          4         120.0        79.0                   저출력    2625   
397  31.0          4         119.0        82.0                   저출력    2720   

     acceleration  model year origin                   car name    kpl  
0            12.0          70    USA  chevrolet chevelle malibu   7.65  
1            11.5          70    USA          buick skylark 320   6.38  
2            11.0          70    USA         plymouth satellite   7.65  
3            12.0          70    USA              amc rebel sst   6.80  
4            10.5          70    USA                ford torino   7.23  
..            ...         ...    ...                        ...    ...  
393          15.6          82    USA            ford mustang gl  11.48  
394          24.6          82     EU                  vw pickup  18.71  
395          11.6          82    USA              dodge rampage  13.60  
396          18.6          82    USA                ford ranger  11.90  
397          19.4          82    USA                 chevy s-10  13.18  

[398 rows x 11 columns]